### PR TITLE
Add suggestion for incorrect field syntax.

### LINKED ISSUE #57684: Misleading compiler hint for wrong keyword constructor syntax
The following code 
```
struct MyClass{
    pub field:i32,
}

fn main(){
    let myClass = MyClass{field=5};
}
```
produces the following error message:
```
error: expected one of `,` or `}`, found `=`
```
This code errors because "field=5" should be "field:5". The error message leads any good python programmer astray. I think the error message should at least mention the possibility of `:`.

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #57684.

This commit adds a suggestion when a `=` character is used when
specifying the value of a field in a struct constructor incorrectly
instead of a `:` character.

r? @estebank

### CODE REVIEW

**Comment by davidtwco** [comment]
Sorry for another addition to your review queue @estebank. Not sure if this is the approach we want to take with this issue or if it is too large a change for the potential gain - if we just want to change to keep the existing error but add `:` (as below) then I can revisit this and do that:

```
error: expected one of `:`, `,` or `}`, found `=` 
```

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -2266,6 +2266,26 @@ impl<'a> Parser<'a> {
...
+                    fieldname.span.shrink_to_hi().to(self.span),
+                    "replace equals symbol with a colon",
+                    ":".to_string(),
+                    Applicability::MachineApplicable,
+                )
+                .emit();
+
+            // Move past the incorrect eq token.
+            self.bump();
+
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Nit: This `if` "arm" can be merged into the previous one, the only different thing it does is the extra error message.

Comment (davidtwco) [reply src/libsyntax/parse/parser.rs]: Oops, well spotted, fixed that.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -2263,8 +2263,24 @@ impl<'a> Parser<'a> {
...
+        let (fieldname, expr, is_shorthand) = if self.look_ahead(1, |t| {
+            t == &token::Colon || t == &token::Eq
+        }) {
             let fieldname = self.parse_field_name()?;
+
+            // Check for an equals token. This means the source incorrectly attempts to
+            // initialize a field with an eq rather than a colon.
+            if self.token == token::Eq {
+                self.diagnostic()
+                    .struct_span_err(self.span, "struct field is initialized with a equals symbol")
```
Comment (ljedrz) [inline src/libsyntax/parse/parser.rs]: Nit: an equals symbol

Comment (davidtwco) [reply src/libsyntax/parse/parser.rs]: Fixed!

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -2263,8 +2263,24 @@ impl<'a> Parser<'a> {
...
+        let (fieldname, expr, is_shorthand) = if self.look_ahead(1, |t| {
+            t == &token::Colon || t == &token::Eq
+        }) {
             let fieldname = self.parse_field_name()?;
+
+            // Check for an equals token. This means the source incorrectly attempts to
+            // initialize a field with an eq rather than a colon.
+            if self.token == token::Eq {
+                self.diagnostic()
+                    .struct_span_err(self.span, "struct field is initialized with an equals symbol")
```
Comment (estebank) [inline src/libsyntax/parse/parser.rs]: Can we reword this?

I feel something closer to the original "expected one of `,` or `}`, found `=`" error would be better, along the lines of "expected one of `:`, found `=`".

Comment (davidtwco) [reply src/libsyntax/parse/parser.rs]: Fixed.

**Comment by davidtwco** [comment]
> 
> 
> Since only ':' is expected, this should just state "expected `:`, found `=`".

I leant toward matching the structure of the existing messages - `expected one of X, found Y` instead of omitting the "one of". I can change this if we prefer otherwise.

**Comment by ljedrz** [comment]
I just feel this might sound like a piece of the error is missing, but it's not a big deal, especially if this is an existing convention for singular items (in which case it might be a better idea to potentially adjust the convention).

**Comment by estebank** [comment]
Yeah, sorry about that it was my typo, my bad. It should be "expected `:`, found `=`", I had originally written "expected one of `:`, `,` or `}`, found `=`" and chose against that to make it even clearer for this case in particular. Can you change it to that?

**Comment by davidtwco** [comment]
@estebank should be fixed now.

**Pull request merged**